artemis-core-client/src/main/java/org/apache/activemq/artemis/core/message/impl/CoreMessage.java

```diff
@@ -579,36 +564,45 @@ public CoreMessage setUserID(UUID userID) {
       try {
+         TypedProperties properties = this.properties;
          if (properties == null) {
-            synchronized (this) {
-               if (properties == null) {
-                  TypedProperties properties = new TypedProperties();
-                  if (buffer != null && propertiesLocation >= 0) {
-                     final ByteBuf byteBuf = buffer.duplicate().readerIndex(propertiesLocation);
-                     properties.decode(byteBuf, coreMessageObjectPools == null ? null : coreMessageObjectPools.getPropertiesDecoderPools());
-                  }
-                  this.properties = properties;
-               }
-            }
+            properties = getOrInitializeTypedProperties();
          }
-
-         return this.properties;
+         return properties;
       } catch (Throwable e) {
-         // This is not an expected error, hence no specific logger created
-         logger.warn("Could not decode properties for CoreMessage[messageID=" + messageID + ",durable=" + durable + ",userID=" + userID + ",priority=" + priority +
-            ", timestamp=" + timestamp + ",expiration=" + expiration + ",address=" + address + ", propertiesLocation=" + propertiesLocation, e);
-         if (buffer != null) {
-            ByteBuf duplicatebuffer = buffer.duplicate();
-            duplicatebuffer.readerIndex(0);
-            logger.warn("Failed message has messageID=" + messageID + " and the following buffer:\n" + ByteBufUtil.prettyHexDump(duplicatebuffer));
-         } else {
-            logger.warn("Failed message has messageID=" + messageID + " and the buffer was null");
-         }
-         throw new RuntimeException(e.getMessage(), e);
+         throw onCheckPropertiesError(e);
+      }
+   }
 
+   private synchronized TypedProperties getOrInitializeTypedProperties() {
+      TypedProperties properties = this.properties;
+      if (properties == null) {
+         properties = new TypedProperties();
+         if (buffer != null && propertiesLocation >= 0) {
+            final ByteBuf byteBuf = buffer.duplicate().readerIndex(propertiesLocation);
+            properties.decode(byteBuf, coreMessageObjectPools == null ? null : coreMessageObjectPools.getPropertiesDecoderPools());
+         }
+         this.properties = properties;
+      }
+      return properties;
+   }
+
+   private RuntimeException onCheckPropertiesError(Throwable e) {
+      // This is not an expected error, hence no specific logger created
+      logger.warn("Could not decode properties for CoreMessage[messageID=" + messageID + ",durable=" + durable + ",userID=" + userID + ",priority=" + priority +
+                     ", timestamp=" + timestamp + ",expiration=" + expiration + ",address=" + address + ", propertiesLocation=" + propertiesLocation, e);
+      final ByteBuf buffer = this.buffer;
+      if (buffer != null) {
+         //risky: a racy modification to buffer indexes could break this duplicate operation
```

Comment:
@michaelandrepearce I would put this error handling into `getOrInitializeTypedProperties` ie synchronized to avoid weird concurrency errors while catching another error, wdyt?

Reply:
the try catch, where on exception we call, onCheckedProperties. Agree this could be moved to getOrInitializeTypedProperties, i would still leave it though where we use a local copy reference.